rules/ast/functions.go

```diff
@@ -498,6 +499,43 @@ func derivImpl(timestamp clientmodel.Timestamp, args []Node) interface{} {
 	return resultVector
 }
 
+// === histogram_quantile(k ScalarNode, vector VectorNode) Vector ===
```

Comment:
From the design doc I expect the following behavior (and I intended to code that):

Given metric 'histo' with job and instance label.

1. `histogram_quantile(.9, rate(histo[5m]))` -> Quantile for each (job, instance) combination.
2. `histogram_quantile(.9, sum by (job, instance, le)(rate(histo[5m])))` -> Same as 1.
3. `histogram_quantile(.9, sum by (instance, le)(rate(histo[5m])))` -> Quantile for each instance.
4. `histogram_quantile(.9, sum by (le)(rate(histo[5m])))` -> One overall quantile.
5. `histogram_quantile(.9, sum (rate(histo[5m])))` -> Empty vector (no `le` label).

If I understand you correctly, you are suggesting the following:

1. `histogram_quantile(.9, rate(histo[5m]))` -> "error out due to clashing labels"
2. `histogram_quantile by (job, instance)(.9, rate(histo[5m]))` -> Quantile for each (job, instance) combination.
3. `histogram_quantile by (instance)(.9, rate(histo[5m]))` -> Quantile for each instance.

This is in general more compact, but why would 1. error out instead of just creating a one-element vector with the overall quantile? What I also like is that the user doesn't have to bother with the `le` label in this case. The downside I see is that you have to specify each label if you don't want to aggregate at all (see 2). The original approach is "purer" as it is following the "do only one thing" semantics. Your new suggestion is probably more useful (if we allow case 1. to work as described), as I expect aggregation to be the usual case (after all, we wanted aggregatable histograms...).